contracts/extensions/ccd006-citycoin-mining-v2.clar

```diff
@@ -0,0 +1,294 @@
+;; Title: CCD006 CityCoin Mining
+;; Version: 2.0.0
+;; Summary: A central city mining contract for the CityCoins protocol.
+;; Description: An extension that provides a mining interface per city, in which each mining participant spends STX per block for a weighted chance to mint new CityCoins per the issuance schedule.
+
+;; TRAITS
+
+(impl-trait .extension-trait.extension-trait)
+(impl-trait .ccd006-trait.ccd006-citycoin-mining-trait)
+
+;; CONSTANTS
+
+(define-constant ERR_UNAUTHORIZED (err u6000))
+(define-constant ERR_INVALID_CITY (err u6001))
+(define-constant ERR_NO_ACTIVATION_DETAILS (err u6002))
+(define-constant ERR_INACTIVE_CITY (err u6003))
+(define-constant ERR_INVALID_USER (err u6004))
+(define-constant ERR_INVALID_TREASURY (err u6005))
+(define-constant ERR_INVALID_DELAY (err u6006))
+(define-constant ERR_INVALID_COMMITS (err u6007))
+(define-constant ERR_NOT_ENOUGH_FUNDS (err u6008))
+(define-constant ERR_ALREADY_MINED (err u6009))
+(define-constant ERR_REWARD_IMMATURE (err u6010))
+(define-constant ERR_NO_VRF_SEED (err u6011))
+(define-constant ERR_DID_NOT_MINE (err u6012))
+(define-constant ERR_NO_MINER_DATA (err u6013))
+(define-constant ERR_ALREADY_CLAIMED (err u6014))
+(define-constant ERR_MINER_NOT_WINNER (err u6015))
+(define-constant ERR_MINING_DISABLED (err u6016))
+
+;; DATA VARS
+
+(define-data-var miningEnabled bool true)
+(define-data-var rewardDelay uint u100)
+
+;; DATA MAPS
+
+(define-map MiningStats
+  { cityId: uint, height: uint }
+  { miners: uint, amount: uint, claimed: bool }
+)
+
+(define-map Miners
+  { cityId: uint, height: uint, userId: uint }
+  { commit: uint, low: uint, high: uint, winner: bool }
+)
+
+(define-map HighValues
+  { cityId: uint, height: uint }
+  uint
+)
+
+(define-map Winners
+  { cityId: uint, height: uint }
+  uint
+)
+
+;; PUBLIC FUNCTIONS
+
+(define-public (is-dao-or-extension)
+  (ok (asserts! (or (is-eq tx-sender .base-dao)
+    (contract-call? .base-dao is-extension contract-caller)) ERR_UNAUTHORIZED
+  ))
+)
+
+(define-public (callback (sender principal) (memo (buff 34)))
+  (ok true)
+)
+
+(define-public (set-reward-delay (delay uint))
+  (begin 
+    (try! (is-dao-or-extension))
+    (print {
+      event: "set-reward-delay",
+      rewardDelay: delay
+    })
+    (asserts! (> delay u0) ERR_INVALID_DELAY)
+    (ok (var-set rewardDelay delay))
+  )
+)
+
+(define-public (set-mining-enabled (status bool))
+  (begin
+    (try! (is-dao-or-extension))
+    (print {
+      event: "set-mining-enabled",
+      miningEnabled: status
+    })
+    (ok (var-set miningEnabled status))
+  )
+)
+
+(define-public (mine (cityName (string-ascii 10)) (amounts (list 200 uint)))
+  (let
+    (
+      (cityId (unwrap! (contract-call? .ccd004-city-registry get-city-id cityName) ERR_INVALID_CITY))
+      (cityInfo (contract-call? .ccd005-city-data get-city-info cityId "mining"))
```

Comment:
Yes, good catch! I'm adding a checklist for remaining changes, and had an idea - this could be a data-var that the DAO can set so it's easier to upgrade in the future.

Reply:
Fixed in [0fee8c6](https://github.com/citycoins/protocol/pull/59/commits/0fee8c6f8a0468192261bace69e6236859561065)